proposals/0000-local-do.rst

```diff
@@ -516,12 +516,26 @@ from different modules if they are imported with the same qualifier:
         g
         return x
 
-Despite of its verbosity when compared to the ``M.do`` approach, the
-``builder.do`` approach is preferred because it groups in a single record
-the operations that work together when desugaring a ``builder.do`` block.
-This single record is easier to import, export and document. In contrast,
-when using ``M.do``, one has to make sure to bring all the needed
-operations into scope.
+The advantages of this approach are that it doesn't need the programmer
+to understand a new notion of expressions having fully settled types.
+Moreover, no type information is necessary to desugar the do notation.
+And lastly, there is no need to write builders: the monadic and applicative
+operations can be brought into scope right from the places where they
+are defined or exported.
+
+A downside of ``M.do`` is that it requires to bring into scope all the
+operations that a ``do`` block needs. In contrast, the builder approach
+only requires to bring a single entity into scope: the builder.
+This single record is easier to import, export and document.
+
+Another downside is that error messages are less specific. Compare
+
+* “Desugaring statement <stmt> requires <field name> but builder <builder expression> doesn't provide it”
+
+with
+
+* “Desugaring statement <stmt> requires ``M.>>`` which is not in scope”
```

Comment:
I don’t see how either prevents us from giving good error messages. There is nothing stopping GHC from making a good error message in either case, .e.g

  The `do` block at <line>:<pos> is qualified by module `M`, but `M` does not export `(>>)` which i
  is needed to handle statement <stmt>.

including things like “You might want to add `(>>)` to the export list of `M`”.

Furthermore I see no problem desugaring `M.do { x : y } = x M.>> do y` even if `M.(>>)` is not in scope, as long as `M` _exports_ `(>>)`. This matches the existing behaviour:
```
Prelude> import Prelude  ()
Prelude> :t (>>)

<interactive>:1:1: error: Variable not in scope: >>
Prelude> foo x = do { x; x}
Prelude> :t foo
foo :: GHC.Base.Monad m => m b -> m b
```

Things are different of course with `-XRebindableSyntax`, and we get an unhelpful error message. But `-XRebindableSyntax` should not be the guiding light here anyways:
```
Prelude> :set -XRebindableSyntax 
Prelude> foo x = do { x; x}

<interactive>:7:14: error: Not in scope: ‘>>’
```

Reply:
I don't think it's very relevant how the message is displayed. The point is that if you are using qualifier-based desugaring, all you can know is that `M.someOperation` is not imported. Maybe you forgot an import statement. Or made a typo in an import statement.

With a record builder, you would know for sure that the builder you are using doesn't support `someOperation`.

From where I stand, it's a much crisper error message.



> Furthermore I see no problem desugaring `M.do { x : y } = x M.>> do y` even if `M.(>>)` is not in scope, as long as `M` _exports_ `(>>)`. This matches the existing behaviour:

I don't know that it makes sense for a namespace to export anything. `M` is not a module, it can even be several module. The Haskell 98 monad syntax is a bit magical in this respect. I'm not sure we could port it to more generic mechanism.